File: src/main/java/ru/uris/Packet.java
package ru.uris;

import ru.DmN.ReflectionUtils;

import java.io.IOException;
import java.util.List;

public class Packet {
    public static int LAST_ID = 0;
    public final int id;
    public final Type type;
    public final boolean request;

    public Packet(int id, Type type, boolean request) {
        this.id = id;
        this.type = type;
        this.request = request;
    }

    public static int nextId() {
        return LAST_ID++;
    }

    public static Packet read(ObjectProviderSocket socket) throws IOException {
        var id = socket.readInt();
        var type = socket.readEnum(Type.class);
        var request = socket.readBoolean();
        return switch (type) {
            case HELLO, CLOSE -> new Packet(id, type, request);
            case OBJECT_LIST -> request ? new PObjectList(id) : new PObjectList(id, socket);
            case METHOD_LIST -> request ? new Packet.PMethodList(id, socket.readInt(), true) : new PMethodList(id, socket);
            case METHOD_CALL -> new PMethodCall(id, socket, request);
        };
    }

    public void write(ObjectProviderSocket socket) throws IOException {
        socket.writeInt(id);
        socket.writeEnum(this.type);
        socket.writeBoolean(this.request);
    }

    public static class PMethodCall extends Packet {
        public RemoteMethod method;
        public Object[] args;
        public Object result;

        public PMethodCall(int pid, RemoteMethod method, Object ... args) {
            super(pid, Type.METHOD_CALL, true);
            this.method = method;
            this.args = args;
            this.result = null;
        }

        public PMethodCall(int pid, ObjectProviderSocket socket, boolean request) throws IOException {
            super(pid, Type.METHOD_CALL, request);
            if (request) {
                var obj = socket.readInt();
                var name = socket.readString();
                var l = socket.readInt();
                var args = new ARType[l];
                for (int i = 0; i < l; i++)
                    args[i] = socket.readARType();
                var ret = socket.readARType();
                this.args = (Object[]) socket.readObject();
                this.method = new RemoteMethod(name, args, ret, obj);
                this.result = null;
            } else {
                this.method = null;
                this.args = null;
                this.result = socket.readObject();
            }
        }

        public PMethodCall(PMethodCall request) {
            super(request.id, Type.METHOD_CALL, false);
            this.method = request.method;
            this.args = request.args;
            this.result = null;
        }

        @Override
        public void write(ObjectProviderSocket socket) throws IOException {
            super.write(socket);
            if (this.request) {
                socket.writeInt(this.method.obj);
                socket.writeString(this.method.name);
                socket.writeInt(this.method.args.length);
                for (var arg : this.method.args)
                    socket.writeARType(arg);
                socket.writeARType(this.method.ret);
                socket.writeObject(this.args);
            } else {
                socket.writeObject(socket.invokeRemoteMethod(this.method, this.args));
            }
        }
    }

    public static class PMethodList extends Packet {
        public final int oid;
        public final RemoteMethod[] methods;

        public PMethodList(int pid, int oid, boolean request) {
            super(pid, Type.METHOD_LIST, request);
            this.oid = oid;
            this.methods = null;
        }

        public PMethodList(int pid, ObjectProviderSocket socket) throws IOException {
            super(pid, Type.METHOD_LIST, false);
            this.oid = socket.readInt();
            var mc = socket.readInt();
            this.methods = new RemoteMethod[mc];
            for (int i = 0; i < mc; i++) {
                var name = socket.readString();
                var ac = socket.readInt();
                var args = new ARType[ac];
                for (int j = 0; j < ac; j++) {
                    args[j] = socket.readARType();
                }
                this.methods[i] = new RemoteMethod(name, args, socket.readARType(),  this.oid);
            }
        }

        @Override
        public void write(ObjectProviderSocket socket) throws IOException {
            super.write(socket);
            socket.writeInt(this.oid);
            if (this.request)
                return;
            var methods = ReflectionUtils.getAllMethods(socket.getObjectPool().get(this.oid).getClass());
            socket.writeInt(methods.length);
            for (var method : methods) {
                socket.writeString(method.getName());
                socket.writeInt(method.getParameterCount());
                for (var arg : method.getParameterTypes()) {
                    socket.writeARType(ARType.of(arg));
                }
                socket.writeARType(ARType.of(method.getReturnType()));
            }
        }
    }

    public static class PObjectList extends Packet {
        public final long[] objects;

        public PObjectList(int id) {
            super(id, Type.OBJECT_LIST, true);
            this.objects = new long[0];
        }

        public PObjectList(int id, List<Object> objects) {
            super(id, Type.OBJECT_LIST, false);
            this.objects = objects.stream().mapToLong(Object::hashCode).toArray();
        }

        public PObjectList(int id, ObjectProviderSocket socket) throws IOException {
            super(id, Type.OBJECT_LIST, false);
            this.objects = new long[socket.readInt()];
            for (int i = 0; i < this.objects.length; i++) {
                this.objects[i] = socket.readLong();
            }
        }

        @Override
        public void write(ObjectProviderSocket socket) throws IOException {
            super.write(socket);
            if (!this.request) {
                socket.writeInt(objects.length);
                for (long object : objects) {
                    socket.writeLong(object);
                }
            }
        }
    }

    public enum Type {
        HELLO,
        CLOSE,

        OBJECT_LIST,
        METHOD_LIST,

        METHOD_CALL
    }
}

File: src/main/kotlin/ru/kuris/Packet.kt
package ru.kuris

import ru.kuris.ARType.Companion.of
import java.io.IOException
import kotlin.reflect.KClass
import kotlin.reflect.full.valueParameters

open class Packet(@JvmField val id: Int, @JvmField val type: Type, @JvmField val request: Boolean) {
    open fun write(socket: ObjectProviderSocket) {
        socket.writeInt(id)
        socket.writeEnum(type)
        socket.writeBoolean(request)
    }

    companion object {
        @JvmField
        var LAST_ID = 0

        @JvmStatic
        fun nextId(): Int = LAST_ID++

        @JvmStatic
        @Throws(IOException::class)
        fun read(socket: ObjectProviderSocket): Packet {
            val id = socket.readInt()
            val type = socket.readEnum(Type::class.java)
            val request = socket.readBoolean()
            return when (type) {
                Type.HELLO, Type.CLOSE -> Packet(id, type, request)
                Type.OBJECT_LIST -> if (request) PObjectList(id) else PObjectList(id, socket)
                Type.METHOD_LIST -> if (request) PMethodList(id, socket.readInt(), true) else PMethodList(id, socket)
            }
        }

        class PMethodList : Packet {
            @JvmField
            val oid: Int

            @JvmField
            val methods: Array<RemoteMethod>

            constructor(pid: Int, oid: Int, request: Boolean) : super(pid, Type.METHOD_LIST, request) {
                this.oid = oid
                this.methods = Array(0) { throw UnsupportedOperationException() }
            }

            constructor(pid: Int, socket: ObjectProviderSocket) : super(pid, Type.METHOD_LIST, false) {
                oid = socket.readInt()
                val mc = socket.readInt()
                val methods = Array<RemoteMethod?>(mc) { null }
                for (i in 0 until mc) {
                    val name = socket.readString()
                    val ac = socket.readInt()
                    val args = Array<ARType?>(ac) { null }
                    for (j in 0 until ac)
                        args[j] = socket.readARType()
                    methods[i] = RemoteMethod(name, args as Array<ARType>, socket.readARType())
                }
                this.methods = methods as Array<RemoteMethod>
            }

            override fun write(socket: ObjectProviderSocket) {
                super.write(socket)
                socket.writeInt(oid)
                if (request)
                    return
                val methods = socket.ObjectPool()[oid]::class.members
                socket.writeInt(methods.size)
                for (method in methods) {
                    socket.writeString(method.name)
                    socket.writeInt(method.valueParameters.size)
                    for (arg in method.valueParameters)
                        socket.writeARType(of(arg.type.classifier as KClass<out Any>))
                    socket.writeARType(of(method.returnType.classifier as KClass<out Any>))
                }
            }
        }

        class PObjectList : Packet {
            @JvmField
            val objects: LongArray

            constructor(id: Int) : super(id, Type.OBJECT_LIST, true) {
                objects = LongArray(0)
            }

            constructor(id: Int, objects: List<Any>) : super(id, Type.OBJECT_LIST, false) {
                this.objects = objects.stream().mapToLong { o -> o.hashCode().toLong() }.toArray()
            }

            constructor(id: Int, socket: ObjectProviderSocket) : super(id, Type.OBJECT_LIST, false) {
                objects = LongArray(socket.readInt())
                for (i in objects.indices)
                    objects[i] = socket.readLong()
            }

            @Throws(IOException::class)
            override fun write(socket: ObjectProviderSocket) {
                super.write(socket)
                if (!request) {
                    socket.writeInt(objects.size)
                    for (obj in objects)
                        socket.writeLong(obj)
                }
            }
        }
    }

    enum class Type {
        HELLO,
        CLOSE,

        OBJECT_LIST,
        METHOD_LIST
    }
}